scipy/optimize/_qap.py

```diff
@@ -0,0 +1,418 @@
+    >>> dist = np.array([[0, 5, 2, 7], [0, 0, 3, 8],
+    ...         [0, 0, 0, 3], [0, 0, 0, 0]])
+    >>> from scipy.optimize import quadratic_assignment
+    >>> res = quadratic_assignment(cost, dist)
+    >>> print(res)
+     col_ind: array([0, 3, 2, 1])
+         nit: 9
+       score: 3260
+
+    To demonstrate explicitly how the `score` value
+    :math:`f(P) = trace(A^T PBP^T )` is calculated, one may construct the
+    permutation matrix, and perform the necessary algebra.
+
+    >>> n = cost.shape[0]
+    >>> P = np.zeros((n, n))
+    >>> P[np.arange(n), res['col_ind']] = 1
+    >>> score = int(np.trace(cost.T @ P @ dist @ P.T))
+    >>> print(score)
+    3260
+
+    As you can see, the value here matches res['score'] reported above.
+    Alternatively, to avoid constructing the permutation matrix, one can also
+    perform the following calculation.
+
+    >>> score = np.trace(cost.T @ dist[np.ix_(res['col_ind'], res['col_ind'])])
+    >>> print(score)
+    3260
+
+    Here, we are simply permuting the distance matrix.
+
+    """
+
+    cost_matrix = np.asarray(cost_matrix)
+    dist_matrix = np.asarray(dist_matrix)
+
+    if seed is None:
+        seed = np.array([[], []]).T
+    seed = np.asarray(seed)
+
+    # ValueError check
+    msg = None
+    if cost_matrix.shape[0] != dist_matrix.shape[0]:
+        msg = "Adjacency matrices must be of equal size"
+    elif (
+        cost_matrix.shape[0] != cost_matrix.shape[1]
+        or dist_matrix.shape[0] != dist_matrix.shape[1]
+    ):
+        msg = "Adjacency matrix entries must be square"
+    elif not (cost_matrix >= 0).all() or not (dist_matrix >= 0).all():
+        msg = "Adjacency matrix entries must be greater than or equal to zero"
+    elif seed.shape[0] > cost_matrix.shape[0]:
+        msg = "There cannot be more seeds than there are nodes"
+    elif seed.shape[1] != 2:
+        msg = "Seed array entry must have two columns"
+    elif not (seed >= 0).all():
+        msg = "Seed array entries must be greater than or equal to zero"
+    elif not (seed <= (cost_matrix.shape[0] - 1)).all():
+        msg = "Seed array entries must be less than the number of nodes"
+    elif not len(set(seed[:, 0])) == len(seed[:, 0]) or not \
+            len(set(seed[:, 1])) == len(seed[:, 1]):
+        msg = "Seed column entries must be unique"
+    elif init_method not in {'barycenter', 'rand'}:
+        msg = "Invalid 'init_method' parameter string"
+    if msg is not None:
+        raise ValueError(msg)
+
+    # TypeError check
+    if type(n_init) is not int or n_init <= 0:
+        msg = "'n_init' must be a positive integer"
+    elif max_iter <= 0 or type(max_iter) is not int:
+        msg = "'max_iter' must be a positive integer"
+    elif type(shuffle_input) is not bool:
+        msg = "'shuffle_input' must be a boolean"
+    elif eps <= 0 or type(eps) is not float:
+        msg = "'eps' must be a positive float"
+    elif type(maximize) is not bool:
```

Comment:
I would remove this check.
Many users will write something like `f(x, maximize=1)` for brevity.